sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -136,19 +135,21 @@ class CacheManager extends Logging {
       spark: SparkSession,
       plan: LogicalPlan,
       cascade: Boolean,
-      blocking: Boolean): Unit = writeLock {
-    val shouldRemove: LogicalPlan => Boolean =
-      if (cascade) {
-        _.find(_.sameResult(plan)).isDefined
-      } else {
-        _.sameResult(plan)
-      }
-    val it = cachedData.iterator()
-    while (it.hasNext) {
-      val cd = it.next()
-      if (shouldRemove(cd.plan)) {
-        cd.cachedRepresentation.cacheBuilder.clearCache(blocking)
-        it.remove()
+      blocking: Boolean): Unit = {
+    writeLock {
+      val shouldRemove: LogicalPlan => Boolean =
+        if (cascade) {
+          _.find(_.sameResult(plan)).isDefined
+        } else {
+          _.sameResult(plan)
+        }
+      val it = cachedData.iterator()
+      while (it.hasNext) {
+        val cd = it.next()
+        if (shouldRemove(cd.plan)) {
+          cd.cachedRepresentation.cacheBuilder.clearCache(blocking)
```

Comment:
we can further reduce the lock, by putting plans to uncache in a list first, and then uncache each plan without lock.

Reply:
I wasn't 100% sure on this change whether the clearCache call was safe to make while the item was still in the cache.  However, if it isn't, I'm not sure this implementation reduces the amount of locking.

Reply:
Ok, in the new implementation clearCache is called after the remove, and outside of a lock.